<!DOCTYPE html>
<html>
<head>
<meta charset="UTF-8">
<title>String-searching-algorithm---Wikipedia-the-free-encyclopedia.html</title></head>
<body>
<h1>String searching algorithm</h1>
<p>In computer science, <b>string searching algorithms</b>, sometimes called <b>string matching algorithms</b>, are an important class of string algorithms that try to find a place where one or several strings (also called patterns) are found within a larger string or text.</p>
<p>Let Σ be an alphabet (finite set). Formally, both the pattern and searched text are vectors of elements of Σ. The Σ may be a usual human alphabet (for example, the letters A through Z in the Latin alphabet). Other applications may use <i>binary alphabet</i> (Σ = {0,1}) or <i>DNA alphabet</i> (Σ = {A,C,G,T}) in bioinformatics.</p>
<p>In practice, how the string is encoded can affect the feasible string search algorithms. In particular if a variable width encoding is in use then it is slow (time proportional to N) to find the Nth character. This will significantly slow down many of the more advanced search algorithms. A possible solution is to search for the sequence of code units instead, but doing so may produce false matches unless the encoding is specifically designed to avoid it.</p>
<p></p>
<h2>Contents</h2>
<ul>
<li>1 Basic classification
<ul>
<li>1.1 Single pattern algorithms</li>
<li>1.2 Algorithms using a finite set of patterns</li>
<li>1.3 Algorithms using an infinite number of patterns</li>
</ul>
</li>
<li>2 Other classification
<ul>
<li>2.1 Naïve string search</li>
<li>2.2 Finite state automaton based search</li>
<li>2.3 Stubs</li>
<li>2.4 Index methods</li>
<li>2.5 Other variants</li>
</ul>
</li>
<li>3 See also</li>
<li>4 Academic conferences on text searching</li>
<li>5 References</li>
<li>6 External links</li>
</ul>
<ul>
<li>1.1 Single pattern algorithms</li>
<li>1.2 Algorithms using a finite set of patterns</li>
<li>1.3 Algorithms using an infinite number of patterns</li>
</ul>
<ul>
<li>2.1 Naïve string search</li>
<li>2.2 Finite state automaton based search</li>
<li>2.3 Stubs</li>
<li>2.4 Index methods</li>
<li>2.5 Other variants</li>
</ul>
<p></p>
<h2>Basic classification</h2>
<p>The various algorithms can be classified by the number of patterns each uses.</p>
<h3>Single pattern algorithms</h3>
<p>Let <i>m</i> be the length of the pattern and let <i>n</i> be the length of the searchable text.</p>
<p>Asymptotic times are expressed using O, Ω, and Θ notation</p>
<p>The <b>Boyer–Moore string search algorithm</b> has been the standard benchmark for the practical string search literature.</p>
<h3>Algorithms using a finite set of patterns</h3>
<ul>
<li>Aho–Corasick string matching algorithm</li>
<li>Commentz-Walter algorithm</li>
<li>Rabin–Karp string search algorithm</li>
</ul>
<h3>Algorithms using an infinite number of patterns</h3>
<p>Naturally, the patterns can not be enumerated in this case. They are represented usually by a regular grammar or regular expression.</p>
<h2>Other classification</h2>
<p>Other classification approaches are possible. One of the most common uses preprocessing as main criteria.</p>
<h3>Naïve string search</h3>
<p>A simple but inefficient way to see where one string occurs inside another is to check each place it could be, one by one, to see if it's there. So first we see if there's a copy of the needle in the first character of the haystack; if not, we look to see if there's a copy of the needle starting at the second character of the haystack; if not, we look starting at the third character, and so forth. In the normal case, we only have to look at one or two characters for each wrong position to see that it is a wrong position, so in the average case, this takes O(<i>n</i> + <i>m</i>) steps, where <i>n</i> is the length of the haystack and <i>m</i> is the length of the needle; but in the worst case, searching for a string like "aaaab" in a string like "aaaaaaaaab", it takes O(<i>nm</i>)</p>
<h3>Finite state automaton based search</h3>
<p>In this approach, we avoid backtracking by constructing a deterministic finite automaton (DFA) that recognizes stored search string. These are expensive to construct—they are usually created using the powerset construction—but are very quick to use. For example,</p>
<h3>Stubs</h3>
<p>Knuth–Morris–Pratt computes a DFA that recognizes inputs with the string to search for as a suffix, Boyer–Moore starts searching from the end of the needle, so it can usually jump ahead a whole needle-length at each step. Baeza–Yates keeps track of whether the previous <i>j</i> characters were a prefix of the search string, and is therefore adaptable to fuzzy string searching. The bitap algorithm is an application of Baeza–Yates' approach.</p>
<h3>Index methods</h3>
<p>Faster search algorithms are based on preprocessing of the text. After building a substring index, for example a suffix tree or suffix array, the occurrences of a pattern can be found quickly. As an example, a suffix tree can be built in <img class="mwe-math-fallback-image-inline tex" alt="\Theta(n)" src="//upload.wikimedia.org/math/5/c/1/5c151486ce3904339687c2bac3cccd8d.png"> time, and all <img class="mwe-math-fallback-image-inline tex" alt="z" src="//upload.wikimedia.org/math/f/b/a/fbade9e36a3f36d3d676c1b808451dd7.png"> occurrences of a pattern can be found in <img class="mwe-math-fallback-image-inline tex" alt="O(m)" src="//upload.wikimedia.org/math/0/e/2/0e2ae329177722b1818828e92b441032.png"> time under the assumption that the alphabet has a constant size and all inner nodes in the suffix tree knows what leafs are underneath them. The latter can be accomplished by running a DFS algorithm from the root of the suffix tree.</p>
<h3>Other variants</h3>
<p>Some search methods, for instance trigram search, are intended to find a "closeness" score between the search string and the text rather than a "match/non-match". These are sometimes called "fuzzy" searches.</p>
<h2>See also</h2>
<ul>
<li>Sequence alignment</li>
<li>Pattern matching</li>
<li>Compressed pattern matching</li>
<li>Approximate string matching</li>
</ul>
<h2>Academic conferences on text searching</h2>
<ul>
<li>Combinatorial pattern matching (CPM), a conference on combinatorial algorithms for strings, sequences, and trees.</li>
<li>String Processing and Information Retrieval (SPIRE), an annual symposium on string processing and information retrieval.</li>
<li>Prague Stringology Conference (PSC), an annual conference on algorithms on strings and sequences.</li>
<li>Competition on Applied Text Searching (CATS), an annual series of evaluations of text searching algorithms.</li>
</ul>
</body>
</html>